yt_dlp/downloader/niconico.py

```diff
@@ -94,47 +103,166 @@ def communicate_ws(reconnect):
+
+            video_id = info_dict['id']
+
+            # Get format index
+            for format_index, fmt in enumerate(info_dict['formats']):
+                if fmt['format_id'] == info_dict['format_id']:
+                    break
+
+            # Get video info
+            total_duration = 0
+            fragment_duration = 0
+            for line in ie._download_webpage(info_dict['url'], video_id, note='Downloading m3u8').splitlines():
+                if '#STREAM-DURATION' in line:
+                    total_duration = int(float(line.split(':')[1]))
+                if '#EXT-X-TARGETDURATION' in line:
+                    fragment_duration = int(line.split(':')[1])
+            if not all({total_duration, fragment_duration}):
+                raise DownloadError('Unable to get required video info')
+
+            ctx = {
+                'filename': filename,
+                'total_frags': math.ceil(total_duration / fragment_duration),
+            }
+
+            self._prepare_and_start_frag_download(ctx, info_dict)
+
+            downloaded_duration = ctx['fragment_index'] * fragment_duration
+            while True:
+                if downloaded_duration > total_duration:
+                    break
+
+                retry_manager = RetryManager(self.params.get('fragment_retries'), self.report_retry)
+                for retry in retry_manager:
+                    try:
+                        # Refresh master m3u8 (if possible) and get the url of the previously-chose format
+                        master_m3u8_url = ws_context._master_m3u8_url()
+                        formats = ie._extract_m3u8_formats(
+                            master_m3u8_url, video_id, query={"start": downloaded_duration}, live=False, note=False, fatal=False)
+                        media_m3u8_url = traverse_obj(formats, (format_index, {dict}, 'url'), get_all=False)
+                        if not media_m3u8_url:
+                            raise DownloadError('Unable to get playlist')
+
+                        # Get all fragments
+                        media_m3u8 = ie._download_webpage(
+                            media_m3u8_url, video_id, note=False, errnote='Unable to download media m3u8')
+                        fragment_urls = traverse_obj(media_m3u8.splitlines(), (
+                            lambda _, v: not v.startswith('#'), {lambda url: urljoin(media_m3u8_url, url)}))
+
+                        with self.DurationLimiter(len(fragment_urls) * fragment_duration * self._PER_FRAGMENT_DOWNLOAD_RATIO):
+                            for fragment_url in fragment_urls:
+                                success = self._download_fragment(ctx, fragment_url, info_dict)
+                                if not success:
+                                    return False
+                                self._append_fragment(ctx, self._read_fragment(ctx))
+                                downloaded_duration += fragment_duration
```

Comment:
What is the expected behavior if user tries to resume a partial download?

Reply:
It continues the download.

`self._prepare_and_start_frag_download` will read the ".ytdl" file and set the `fragment_index` field, then `downloaded_duration` will be calculated. The server will return fragments starting from `downloaded_duration` because of `query={"start": downloaded_duration}` .